runtime/mbarrier.cu

```diff
@@ -0,0 +1,67 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+// Reference:
+// https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#asynchronous-barrier
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier
+// https://github.com/NVIDIA/cutlass/blob/main/include/cute/arch/copy_sm90_desc.hpp
+
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 800))
+
+namespace mbarrier {
+
+__device__ inline void init(
+    uint32_t smem_barrier_ptr,
+    uint32_t thread_count = 1) {
+  asm volatile(
+      "mbarrier.init.shared.b64 [%0], %1;\n" ::"r"(smem_barrier_ptr),
+      "r"(thread_count));
+}
+
+__device__ inline void inval(uint32_t smem_barrier_ptr) {
+  asm volatile("mbarrier.inval.shared.b64 [%0];\n" ::"r"(smem_barrier_ptr));
+}
+
+__device__ inline uint64_t arrive(uint32_t smem_barrier_ptr) {
+  volatile uint64_t state;
+  asm volatile("mbarrier.arrive.shared.b64 %0, [%1];\n"
+               : "=l"(state)
+               : "r"(smem_barrier_ptr));
+  return state;
+}
+
+__device__ inline void wait(uint32_t smem_barrier_ptr, uint64_t state) {
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
```

Comment:
The [sample](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-arrive:~:text=waitLoop%3A%0Ambarrier.try_wait.shared.b64%20%20%20%20complete%2C%20%5BshMem%5D%2C%20%25r1%3B%0A%40!complete%20bra%20waitLoop%3B) code from the doc seems simpler:

Reply:
Yeah, agree. Changed that.